src/util/newconfig/ConfigConstraints.hpp

```diff
@@ -0,0 +1,401 @@
+
+//------------------------------------------------------------------------------
+/*
+   This file is part of clio: https://github.com/XRPLF/clio
+   Copyright (c) 2024, the clio developers.
+
+   Permission to use, copy, modify, and distribute this software for any
+   purpose with or without fee is hereby granted, provided that the above
+   copyright notice and this permission notice appear in all copies.
+
+   THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+   WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+   MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+   ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+   WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+   ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include "util/newconfig/Errors.hpp"
+
+#include <array>
+#include <cstdint>
+#include <limits>
+#include <optional>
+#include <string>
+#include <variant>
+
+namespace util::config {
+class ValueView;
+class ConfigValue;
+
+/**
+ * @brief specific values that are accepted for channel names in config
+ */
+static constexpr std::array<char const*, 7> channels = {
+    "General",
+    "WebServer",
+    "Backend",
+    "RPC",
+    "ETL",
+    "Subscriptions",
+    "Performance",
+};
+
+/**
+ * @brief specific values that are accepted for logger levels in config
+ */
+static constexpr std::array<char const*, 7> logLevels = {
+    "trace",
+    "debug",
+    "info",
+    "warning",
+    "error",
+    "fatal",
+    "count",
+};
+
+/**
+ * @brief specific values that are accepted for logger tag style in config
+ */
+static constexpr std::array<char const*, 5> logTags = {
+    "int",
+    "uint",
+    "null",
+    "none",
+    "uuid",
+};
+
+/**
+ * @brief Minimum API version supported by this build
+ *
+ * Note: Clio does not natively support v1 and only supports v2 or newer.
+ * However, Clio will forward all v1 requests to rippled for backward compatibility.
+ */
+static constexpr uint32_t API_VERSION_MIN = 1u;
+
+/**
+ * @brief Maximum API version supported by this build
+ */
+static constexpr uint32_t API_VERSION_MAX = 2u;
+
+/**
+ * @brief A interface to enforce constraints on certain values within ClioConfigDefinition.
+ */
+class Constraint {
+public:
+    using ValueType = std::variant<int64_t, std::string, bool, double>;
```

Comment:
This duplicates `Type` from `ConfigValue`. Probably it would be better to move this definition to a separate header file and reuse everywhere.